File: ukpsummarizer-be/cplex/cpoptimizer/examples/src/cpp/talent.cpp
// -------------------------------------------------------------- -*- C++ -*-
// File: ./examples/src/cpp/talent.cpp
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
//
// 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5725-A06 5725-A29
// Copyright IBM Corporation 1990, 2014. All Rights Reserved.
//
// Note to U.S. Government Users Restricted Rights:
// Use, duplication or disclosure restricted by GSA ADP Schedule
// Contract with IBM Corp.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

This example is inspired from the talent hold cost scheduling problem
described in:

T.C.E Cheng, J. Diamond, B.M.T. Lin.  Optimal scheduling in film
production to minimize talent holding cost.  Journal of Optimization
Theory and Applications, 79:197-206, 1993.

of which the 'Rehearsal problem' is a specific case:

Barbara M. Smith.  Constraint Programming In Practice: Scheduling
                   a Rehearsal.  Report APES-67-2003, September 2003.

See: http://www.csplib.org/Problems/prob039/


------------------------------------------------------------ */

#include <ilcp/cp.h>

// Read talent scheduling problem data from a file
IloBool ReadData(const char * filename,
                 IloIntArray actorPay,
                 IloIntArray sceneDuration,
                 IloArray<IloIntSet> actorInScene) {
  IloEnv env = actorPay.getEnv();
  std::ifstream in(filename);
  if (!in.good())
    return IloFalse;

  IloInt numActors, numScenes, a, s;

  in >> numActors;
  for (a = 0; a < numActors; a++) {
    IloInt pay;
    in >> pay;
    actorPay.add(pay);
  }

  in >> numScenes;
  for (s = 0; s < numScenes; s++) {
    IloInt duration;
    in >> duration;
    sceneDuration.add(duration);
  }

  for (a = 0; a < numActors; a++) {
    actorInScene.add(IloIntSet(env));
    for (s = 0; s < numScenes; s++) {
      IloBool inScene;
      in >> inScene;
      if (inScene)
        actorInScene[a].add(s);
    }
  }
  if (!in.good())
    return IloFalse;

  in.close();
  return IloTrue;
}

// Build the talent scheduling model
IloModel BuildModel(IloIntVarArray scene,
                    IloIntExpr& idleCost,
                    IloIntArray actorCost,
                    IloIntArray sceneDuration,
                    IloArray<IloIntSet> actorInScene) {
  IloEnv env = scene.getEnv();
  IloInt numScenes = scene.getSize();
  IloInt numActors = actorCost.getSize();
  IloModel model(env);

  // Make the slot-based secondary model
  IloIntVarArray slot(env, numScenes, 0, numScenes - 1);
  model.add(IloInverse(env, scene, slot));

  // Expression representing the global cost
  idleCost = IloIntExpr(env, 0);

  // Loop over all actors, building cost
  for (IloInt a = 0; a < numActors; a++) {
    // Expression for the waiting time for this actor
    IloIntExpr actorWait(env);

    // Calculate the first and last slots where this actor plays
    IloIntVarArray position(env);
    for (IloIntSet::Iterator it(actorInScene[a]); it.ok(); ++it)
      position.add(slot[*it]);
    IloIntExpr firstSlot = IloMin(position);
    IloIntExpr lastSlot = IloMax(position);

    // If an actor is not in a scene,
    // he waits if he is on set when the scene is filmed
    for (IloInt s = 0; s < numScenes; s++) {
      if (!actorInScene[a].contains(s)) { // not in scene
        IloIntExpr wait = (firstSlot <= slot[s] && slot[s] <= lastSlot);
        actorWait += sceneDuration[s] * wait;
      }
    }

    // Accumulate the cost of waiting time for this actor
    idleCost += actorCost[a] * actorWait;
  }
  model.add(IloMinimize(env, idleCost));
  return model;
}

int main(int argc, const char * argv[]) {
  IloEnv env;
  try {
    const char * inputFile = "../../../examples/data/rehearsal.data";
    IloNum tlim = 10.0;
    if (argc > 1)
      inputFile = argv[1];
    if (argc > 2)
      tlim = atof(argv[2]);

    IloIntArray actorPay(env);
    IloIntArray sceneDuration(env);
    IloArray<IloIntSet> actorInScene(env);
    IloBool ok = ReadData(inputFile, actorPay, sceneDuration, actorInScene);
    if (!ok) {
      env.out() << "Error reading " << inputFile << std::endl;
    }
    else {
      // Create the decision variables, cost, and the model
      IloInt numScenes = sceneDuration.getSize();
      IloInt numActors = actorPay.getSize();
      IloIntVarArray scene(env, numScenes, 0, numScenes - 1);
      IloIntExpr idleCost;
      IloModel model = BuildModel(
        scene, idleCost, actorPay, sceneDuration, actorInScene
      );

      // Create the CP solver
      IloCP cp(model);
      cp.setParameter(IloCP::TimeLimit, tlim);
      cp.solve();

      cp.out() << "Solution of idle cost " << cp.getValue(idleCost) << std::endl;
      cp.out() << "Order:";
      for (IloInt s = 0; s < numScenes; s++)
        cp.out() << " " << 1 + cp.getValue(scene[s]);
      cp.out() << std::endl;

      // Give more detailed information on the schedule
      for (IloInt a = 0; a < numActors; a++) {
        cp.out() << "|";
        for (IloInt s = 0; s < numScenes; s++) {
          IloInt sc = cp.getValue(scene[s]);
          for (IloInt d = 0; d < sceneDuration[sc]; d++) {
            if (actorInScene[a].contains(sc))
              cp.out() << "X";
            else
              cp.out() << ".";
          }
          cp.out() << "|";
        }
        cp.out() << "  (Rate = " << actorPay[a] << ")" << std::endl;
      }
      cp.end();
    }
  } catch (IloException & ex) {
    env.out() << "Caught: " << ex << std::endl;
  }
  env.end();
  return 0;
}

File: ukpsummarizer-be/cplex/cpoptimizer/examples/src/java/Talent.java
// ---------------------------------------------------------------*- Java -*-
// File: ./examples/src/java/Talent.java
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
//
// 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5725-A06 5725-A29
// Copyright IBM Corporation 1990, 2017. All Rights Reserved.
//
// Note to U.S. Government Users Restricted Rights:
// Use, duplication or disclosure restricted by GSA ADP Schedule
// Contract with IBM Corp.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

This example is inspired from the talent hold cost scheduling problem
described in:

T.C.E Cheng, J. Diamond, B.M.T. Lin.  Optimal scheduling in film
production to minimize talent holding cost.  Journal of Optimization
Theory and Applications, 79:197-206, 1993.

of which the 'Rehearsal problem' is a specific case:

Barbara M. Smith.  Constraint Programming In Practice: Scheduling
                   a Rehearsal.  Report APES-67-2003, September 2003.

See: http://www.csplib.org/Problems/prob039/

------------------------------------------------------------ */

import ilog.concert.*;
import ilog.cp.*;

import java.io.*;
import java.util.Iterator;
import java.util.Vector;

public class Talent extends IloCP {

  public static IloCP cp;

  private int numActors, numScenes;
  private int[] actorPay, sceneDuration;
  private IloIntSet[] actorInScene;
  private IloIntExpr idleCost;
  private IloIntVar scene[];

  public Talent(String fileName) throws IOException, IloException {

    cp = new IloCP();
    this.readData(fileName);
    this.buildModel();
  }

  private static class DataReader {

    private StreamTokenizer st;

    public DataReader(String filename) throws IOException {
      FileInputStream fstream = new FileInputStream(filename);
      Reader r = new BufferedReader(new InputStreamReader(fstream));
      st = new StreamTokenizer(r);
    }

    public int next() throws IOException {
      st.nextToken();
      return (int) st.nval;
    }
  }

  private void readData(String fileName) throws IOException, IloException {

    DataReader data = new DataReader(fileName);

    numActors = data.next();
    actorPay = new int[numActors];
    for (int a = 0; a < numActors; a++)
      actorPay[a] = data.next();

    numScenes = data.next();
    sceneDuration = new int[numScenes];
    for (int s = 0; s < numScenes; s++)
      sceneDuration[s] = data.next();

    actorInScene = new IloIntSet[numActors];
    for (int a = 0; a < numActors; a++) {
      int[] inScene = new int[numScenes];
      int nbScene = 0;
      for (int s = 0; s < numScenes; s++) {
        inScene[s] = data.next();
        if (inScene[s] != 0)
          nbScene++;
      }
      int[] playScene = new int[nbScene];
      int n = 0;
      for (int s = 0; s < numScenes; s++) {
        if (inScene[s] != 0) {
          playScene[n] = s;
          n++;
        }
      }
      actorInScene[a] = cp.intSet(playScene);
    }
  }

  private void buildModel() throws IloException {

    // Create the decision variables, cost, and the model
    scene = new IloIntVar[numScenes];
    for (int s = 0; s < numScenes; s++)
      scene[s] = cp.intVar(0, numScenes - 1);

    // Expression representing the global cost
    idleCost = cp.intExpr();

    // Make the slot-based secondary model
    IloIntVar[] slot = new IloIntVar[numScenes];
    for (int s = 0; s < numScenes; s++)
      slot[s] = cp.intVar(0, numScenes - 1);
    cp.add(cp.inverse(scene, slot));

    // Loop over all actors, building cost
    for (int a = 0; a < numActors; a++) {
      // Expression for the waiting time for this actor
      IloIntExpr actorWait = cp.intExpr();

      // Calculate the first and last slots where this actor plays
      Vector<IloIntVar> position = new Vector<IloIntVar>();

      Iterator it = actorInScene[a].iterator();
      while (it.hasNext())
        position.add(slot[(Integer)it.next()]);

      IloIntExpr firstSlot = cp.min(position.toArray(new IloIntVar[position.size()]));
      IloIntExpr lastSlot = cp.max(position.toArray(new IloIntVar[position.size()]));

      // If an actor is not in a scene, he waits
      // if he is on set when the scene is filmed
      for (int s = 0; s < numScenes; s++) {
        if (!actorInScene[a].contains(s)) { // not in scene
          IloIntExpr wait = cp.and(cp.le(firstSlot, slot[s]), cp.le(
              slot[s], lastSlot));
          actorWait = cp.sum(actorWait, cp.prod(sceneDuration[s],
              wait));
        }
      }

      // Accumulate the cost of waiting time for this actor
      idleCost = cp.sum(idleCost, cp.prod(actorPay[a], actorWait));
    }
    cp.add(cp.minimize(idleCost));
  }

  private void display() {

    System.out.println("Solution of idle cost " + (int) cp.getValue(idleCost));
    System.out.print("Order:");
    for (int s = 0; s < numScenes; s++)
      System.out.print(" " + ((int) cp.getValue(scene[s]) + 1));
    System.out.println();

    // Give more detailed information on the schedule
    for (int a = 0; a < numActors; a++) {
      System.out.print("|");
      for (int s = 0; s < numScenes; s++) {
        int sc = (int) cp.getValue(scene[s]);
        for (int d = 0; d < sceneDuration[sc]; d++) {
          if (actorInScene[a].contains(sc))
            System.out.print("X");
          else
            System.out.print(".");
        }
        System.out.print("|");
      }
      System.out.println("  Rate = " + actorPay[a] + ")");
    }
  }

  public static void main(String args[]) throws IOException, IloException {

    String inputFile = "../../../examples/data/rehearsal.data";
    if (args.length > 1)
      inputFile = args[1];

    Talent talent = new Talent(inputFile);
    cp.solve();

    talent.display();
  }
}
